frontend/src/components/dashboard/hooks/useAutomaticGrouping.ts

```diff
@@ -0,0 +1,322 @@
+        )
+        .with(
+          { statementType: StatementNodeType.sequence },
+          (node) => node.statements,
+        )
+        .with(
+          { statementType: StatementNodeType.variableDeclaration },
+          (node) => [node.value],
+        )
+        .exhaustive(),
+    )
+    .with({ nodeType: AstNodeType.expression }, (node) =>
+      match(node)
+        .returnType<AstNode[]>()
+        .with(
+          { expressionType: ExpressionNodeType.functionCall },
+          (node) => node.arguments,
+        )
+        .with({ expressionType: ExpressionNodeType.literal }, () => [])
+        .with(
+          { expressionType: ExpressionNodeType.operator },
+          (node) => node.operands,
+        )
+        .with({ expressionType: ExpressionNodeType.variable }, () => [])
+        .exhaustive(),
+    )
+    .exhaustive();
+
+const getInputForNode = (
+  a: AstNode,
+): {
+  root: AstNode;
+  lfn: (node: AstNode) => string;
+  cfn: (node: AstNode) => AstNode[];
+} => ({
+  root: a,
+  lfn,
+  cfn,
+});
+
+const computePqDistance = (
+  a: AstNode,
+  b: AstNode,
+  p = 2,
+  q = 3,
+): Promise<number> =>
+  new Promise((resolve) =>
+    jqgram.distance(
+      getInputForNode(a),
+      getInputForNode(b),
+      { p, q },
+      (result) => resolve(result.distance),
+    ),
+  );
+
+interface IntermediateGroup {
+  solutions: CurrentAnalysis[];
+}
+
+const meanDistance = (
+  a: IntermediateGroup,
+  b: IntermediateGroup,
+  distanceMap: Map<number, Map<number, number>>,
+): number => {
+  let sum = 0;
+
+  for (const solutionA of a.solutions) {
+    for (const solutionB of b.solutions) {
+      sum +=
+        distanceMap.get(solutionA.solutionId)?.get(solutionB.solutionId) ?? 0;
+    }
+  }
+
+  return sum / (a.solutions.length * b.solutions.length);
+};
+
+export const useAutomaticGrouping = (
+  isAutomaticGrouping: boolean,
+  numberOfGroups: number,
+  solutionsInput: CurrentAnalysis[] | undefined,
+  filters: FilterCriterion[],
+  xAxis: AxesCriterionType,
+  yAxis: AxesCriterionType,
+): (() => Promise<{
+  groupAssignments: SolutionGroupAssignment[];
+  automaticGroups: AutomaticGroup[];
+}>) =>
+  useCallback(async () => {
+    const solutions = solutionsInput?.filter((solution) =>
+      filters.every((filter) => matchesFilter(filter, solution)),
+    );
+
+    if (!solutions || solutions.length === 0 || !isAutomaticGrouping) {
+      return { groupAssignments: [], automaticGroups: [] };
+    }
+
+    let groups: IntermediateGroup[] = solutions.map((solution) => ({
+      solutions: [solution],
+    }));
+
+    const distances: Map<number, Map<number, number>> = new Map();
+    for (const solution of solutions) {
+      distances.set(solution.solutionId, new Map());
+    }
+
+    const promises: Promise<void>[] = [];
+
+    for (let i = 0; i < solutions.length; i++) {
+      for (let j = i + 1; j < solutions.length; j++) {
+        promises.push(
+          computePqDistance(
```

Comment:
Not needed for now, but if this becomes its own tiny distance-computation library, we could consider making it generic and call `computDistance("pq", ...` for the time being, with "pq" being literally the only accepted value 😆